README.md
# movehub async

[![npm version](https://badge.fury.io/js/movehub-async.svg)](https://badge.fury.io/js/movehub-async)

Simple to use asynchronous methods for the [Lego Boost Move Hub](https://github.com/hobbyquaker/node-movehub)

## Setup

* Install [Noble prerequisites](https://github.com/noble/noble#prerequisites)

* Install movehub-async

```sh
$ npm install movehub-async
```

## Usage

```js
const boost = require('movehub-async');

const hub = await boost.getHubAsync();

// Turn light from red to green
await hub.ledAsync('red');
await hub.ledAsync('yellow');
await hub.ledAsync('green');

// Turn A & B motors for 10 seconds with power 20
await hub.motorTimeMultiAsync(10, 20, 20);
// Turn motor C 600 degrees with power 5
await hub.motorAngleAsync('C', 600, 5);
```

It is also possible to wait that motor execution has stopped

```js
await hub.ledAsync('red');
// Continue when led is red
await hub.motorTimeMultiAsync(10, 20, 20, true);
// Continue 10 sec later
await hub.motorTimeMultiAsync(5, 20, 20, true);
// Continue 5 sec later
await hub.motorAngleAsync('C', 800, 50, true);
// Continue some time later
await hub.ledAsync('green');
// Continue when led is green
```

Package contains also simple methods to drive for a specified distance and turn a specified angle. By default drive and turn methods will wait the execution has stopped.

```js
// Drive 2 meters forward
await hub.drive(200);
// After 2 meter drive, turn 90 degrees to the right
await hub.turn(90);
// Drive 1 meter backwards
await hub.drive(-100);
// Turn 180 degrees to the left
await hub.turn(-180); 
```

## API

Check complete non-async API definition from [Lego Boost Move Hub](https://github.com/hobbyquaker/node-movehub). Asynchronous version of the method has an _Async_-suffix in the name, e.g. `motorTimeMulti` -> `motorTimeMultiAsync`.

## Boost

### boost.getHubAsync()

Create a connection to the Hub. Internally calls `bleReadyAsync`, `hubFoundAsync` and `connectAsync`.

```js
const hub = await boost.getHubAsync();
```

### boost.bleReadyAsync()

Wait for BLE device to be ready.

```js
await boost.bleReadyAsync();
```

### boost.hubFoundAsync()

Wait for MoveHub found event.

```js
const connectDetails = await boost.hubFoundAsync();
```

### boost.connectAsync(connectDetails)

Initialize and wait for the connection to the Hub.

```js
const hub = await boost.connectAsync(connectDetails);
```

## Hub

### hub.ledAsync(color)

Control the LED on the Move Hub.

```js
await hub.ledAsync('red');
```

### hub.motorTimeAsync(port, seconds, dutyCycle = 100, wait = false)

Run a motor for specific time. Await returns when command is sent to Hub.

```js
await hub.motorTimeAsync('C', 5, 50);
// Continue almost immediately when command is sent to Hub

await hub.motorTimeAsync('C', 5, 50, true);
// Continue 5 seconds later
```

### hub.motorTimeMultiAsync(seconds, dutyCycleA = 100, dutyCycleB = 100, wait = false)

Run both motors (A and B) for specific time. Await returns when command is sent to Hub.

```js
// Drive forward for 10 seconds
await hub.motorTimeMultiAsync(10, 20, 20, true);
// Continue 10 seconds later
```

### hub.motorAngleAsync(port, angle, dutyCycle = 100, wait = false)

Turn a motor by specific angle. Await returns when command is sent to Hub.

```js
// Turn ~180 degrees  
await hub.motorAngleAsync('B', 980, 100, true);
// Continue after the turn
```

### hub.motorAngleMultiAsync(angle, dutyCycleA = 100, dutyCycleB = 100, wait = false)

Turn both motors (A and B) by specific angle. Await returns when command is sent to Hub.

```js
// Drive forward
await hub.motorAngleMultiAsync(500, 100, 100);
// Continue immediately after command is sent to Hub
```

### hub.drive(centimeters, wait = true)

Drive specified distance. By default drive-method's return promise will resolve when the distance has been driven.

__Note:__ Drive method is implemented with Lego Boost Vernie

```js
// Drive forward 2 meters
await hub.drive(200);
// Continue after drive is finished
```

### hub.useMetricUnits()

Use metric untis in drive-method. Metric is default.

```js
// Drive forward 200 cm
await hub.drive(200);

hub.useImperialUnits();

// Drive forward 200 inches
await hub.drive(200);
```

### hub.useImperialUnits()

Use imperial units with drive-method.

### hub.setFrictionModifier(modifier)

If drive method's distance is not correct, friction modifier can be changed.

```js
// Drive forward 100cm
await hub.drive(100);

// Distance was only 90cm, update modifier
hub.setFrictionModifier(1.1);

// Drive 100cm
await hub.drive(100);
```

### hub.turn(degrees, wait = true)

Turn specified angle to either right (positive number) or left (negative number). By default turn-method's promise will resolve when the angle has been turned.

__Note:__ Turn method is implemented with Lego Boost Vernie

```js
const hub = await boost.getHubAsync();
// Drive 1 meter square
await hub.drive(100);
await hub.turn(90);
await hub.drive(100);
await hub.turn(90);
await hub.drive(100);
await hub.turn(90);
await hub.drive(100);
await hub.turn(90);
```

### hub.driveUntil(distance = 0, wait = true)

Drive until the sensor shows an object in defined distance. The distance sensor is not very sensitive or accurate. By default the bot will stop when the sensor notices a wall for the first time. Sensor distance values are usualy between 110-50.

```js
await hub.driveUntil();
```

### hub.turnUntil(direction = 1, wait = true)

Turn until sensor doesn't detect any blocking object. 1 or any positive number turns to the right (default) and 0 or any negative number turns to the left.

```js
// Turn to the right
await hub.turnUntil();
// Turn to the right
await hub.turnUntil(1);
// Turn to the left
await hub.turnUntil(0);
```

## Example project

[lego-boost-ai](https://github.com/ttu/lego-boost-ai) has a simple AI and manual controls for Lego Boost.

## Tester

[tester.js](https://github.com/ttu/node-movehub-async/blob/master/tester.js)

## Use development version

It is possible to use development version from GitHub, which may contain unreleased features.

```sh
$ npm install git+https://git@github.com/ttu/node-movehub-async.git
```

## Changelog

[Changelog](https://github.com/ttu/node-movehub-async/blob/master/CHANGELOG.md)

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

## License

Licensed under the [MIT](https://github.com/ttu/node-movehub-async/blob/master/LICENSE) License.

CHANGELOG.md
# Changelog
 
## [Unreleased]

## [0.5.0] - 2018-09-30

- Drive until method
- Turn until method

## [0.4.0] - 2018-04-07

- Add drive method for driving specified distance
- Add turn method for turning specified angle

## [0.3.3] - 2018-03-17

- Add dutyCycle default values to motor methods
- Update JSDocs

## [0.3.2] - 2018-03-16

- npm README update

## [0.3.1] - 2018-03-16

- Fix motorAngle wait
- Update JSDocs

## [0.3.0] - 2018-03-16

- Asynchronous motor methods

## [0.2.0] - 2018-03-12

- Simple getHubAsync connection method

## [0.1.1] - 2018-03-11
 
- JSDoc comments

## [0.1.0] - 2018-03-10
 
- Functions for connection creation
- Led control method

movehub-async.js
const { Boost, Hub } = require('movehub/movehub');

const CALLBACK_TIMEOUT_MS = 1000 / 3;
const METRIC_MODIFIER = 28.5;
const IMPERIAL_MODIFIER = METRIC_MODIFIER / 4;
const TURN_MODIFIER = 2.56;
const DRIVE_SPEED = 25;
const TURN_SPEED = 20;
const DEFAULT_STOP_DISTANCE = 105;
const DEFAULT_CLEAR_DISTANCE = 120;

const waitForValueToSet = function(valueName, compareFunc = (valueName) => this[valueName], timeoutMs = 0) {
  if (compareFunc.bind(this)(valueName)) return Promise.resolve(this[valueName]);

  return new Promise((resolve, reject) => {
    setTimeout(async () => resolve(await waitForValueToSet.bind(this)(valueName, compareFunc, timeoutMs)), timeoutMs + 100);
  });
};

/**
 * Disconnect Hub
 * @method Hub#disconnectAsync
 * @returns {Promise<boolean>} disconnection successful
 */
Hub.prototype.disconnectAsync = function() {
  this.disconnect();
  return waitForValueToSet.bind(this)('hubDisconnected');
};

/**
 * Execute this method after new instance of Hub is created
 * @method Hub#afterInitialization
 */
Hub.prototype.afterInitialization = function() {
  this.hubDisconnected = null;
  this.ports = {
    A: { angle: 0 },
    B: { angle: 0 },
    AB: { angle: 0 },
    C: { angle: 0 },
    D: { angle: 0 },
    LED: { angle: 0 }
  };
  this.useMetric = true;
  this.modifier = 1;

  this.on('rotation', rotation => this.ports[rotation.port].angle = rotation.angle);
  this.on('disconnect', () => this.hubDisconnected = true);
  this.on('distance', distance => this.distance = distance);
};

/**
 * Control the LED on the Move Hub
 * @method Hub#ledAsync
 * @param {boolean|number|string} color
 * If set to boolean `false` the LED is switched off, if set to `true` the LED will be white.
 * Possible string values: `off`, `pink`, `purple`, `blue`, `lightblue`, `cyan`, `green`, `yellow`, `orange`, `red`,
 * `white`
 * @returns {Promise}
 */
Hub.prototype.ledAsync = function(color) {
  return new Promise((resolve, reject) => {
    this.led(color, () => {
      // Callback is executed when command is sent and it will take some time before MoveHub executes the command
      setTimeout(resolve, CALLBACK_TIMEOUT_MS);
    });
  });
}

/**
 * Run a motor for specific time
 * @method Hub#motorTimeAsync
 * @param {string|number} port possible string values: `A`, `B`, `AB`, `C`, `D`.
 * @param {number} seconds
 * @param {number} [dutyCycle=100] motor power percentage from `-100` to `100`. If a negative value is given rotation
 * is counterclockwise.
 * @param {boolean} [wait=false] will promise wait unitll motorTime run time has elapsed
 * @returns {Promise}
 */
Hub.prototype.motorTimeAsync = function(port, seconds, dutyCycle = 100, wait = false) {
  return new Promise((resolve, reject) => {
    this.motorTime(port, seconds, dutyCycle, () => {
      setTimeout(resolve, wait ? CALLBACK_TIMEOUT_MS + seconds * 1000 : CALLBACK_TIMEOUT_MS);
    });
  });
}

/**
 * Run both motors (A and B) for specific time
 * @method Hub#motorTimeMultiAsync
 * @param {number} seconds
 * @param {number} [dutyCycleA=100] motor power percentage from `-100` to `100`. If a negative value is given rotation
 * is counterclockwise.
 * @param {number} [dutyCycleB=100] motor power percentage from `-100` to `100`. If a negative value is given rotation
 * is counterclockwise.
 * @param {boolean} [wait=false] will promise wait unitll motorTime run time has elapsed
 * @returns {Promise}
 */
Hub.prototype.motorTimeMultiAsync = function(seconds, dutyCycleA = 100, dutyCycleB = 100, wait = false) {
  return new Promise((resolve, reject) => {
    this.motorTimeMulti(seconds, dutyCycleA, dutyCycleB, () => {
      setTimeout(resolve, wait ? CALLBACK_TIMEOUT_MS + seconds * 1000 : CALLBACK_TIMEOUT_MS);
    });
  });
}

/**
 * Turn a motor by specific angle
 * @method Hub#motorAngleAsync
 * @param {string|number} port possible string values: `A`, `B`, `AB`, `C`, `D`.
 * @param {number} angle - degrees to turn from `0` to `2147483647`
 * @param {number} [dutyCycle=100] motor power percentage from `-100` to `100`. If a negative value is given
 * rotation is counterclockwise.
 * @param {boolean} [wait=false] will promise wait unitll motorAngle has turned
 * @returns {Promise}
 */
Hub.prototype.motorAngleAsync = function(port, angle, dutyCycle = 100, wait = false) {
  return new Promise((resolve, reject) => {
    this.motorAngle(port, angle, dutyCycle, async () => {
      if (wait) {
        let beforeTurn;
        do {
          beforeTurn = this.ports[port].angle;
          await new Promise(res => setTimeout(res, CALLBACK_TIMEOUT_MS))
        } while(this.ports[port].angle !== beforeTurn)
        resolve();
      } else {
        setTimeout(resolve, CALLBACK_TIMEOUT_MS);
      }
    });
  });
}

/**
 * Turn both motors (A and B) by specific angle
 * @method Hub#motorAngleMultiAsync
 * @param {number} angle degrees to turn from `0` to `2147483647`
 * @param {number} [dutyCycleA=100] motor power percentage from `-100` to `100`. If a negative value is given
 * rotation is counterclockwise.
 * @param {number} [dutyCycleB=100] motor power percentage from `-100` to `100`. If a negative value is given
 * rotation is counterclockwise.
 * @param {boolean} [wait=false] will promise wait unitll motorAngle has turned
 * @returns {Promise}
 */
Hub.prototype.motorAngleMultiAsync = function(angle, dutyCycleA = 100, dutyCycleB = 100, wait = false) {
  return new Promise((resolve, reject) => {
    this.motorAngleMulti(angle, dutyCycleA, dutyCycleB, async () => {
      if (wait) {
        let beforeTurn;
        do {
          beforeTurn = this.ports['AB'].angle;
          await new Promise(res => setTimeout(res, CALLBACK_TIMEOUT_MS))
        } while(this.ports['AB'].angle !== beforeTurn)
        resolve();
      } else {
        setTimeout(resolve, CALLBACK_TIMEOUT_MS);
      }
    });
  });
}

/**
 * Use metric units (default)
 * @method Hub#useMetricUnits
 */
Hub.prototype.useMetricUnits = function() {
  this.useMetric = true;
}

/**
 * Use imperial units
 * @method Hub#useImperialUnits
 */
Hub.prototype.useImperialUnits = function() {
  this.useMetric = false;
}

/**
 * Set friction modifier
 * @method Hub#setFrictionModifier
 * @param {number} modifier friction modifier
 */
Hub.prototype.setFrictionModifier = function(modifier) {
  this.modifier = modifier;
}

/**
 * Drive specified distance
 * @method Hub#drive 
 * @param {number} distance distance in centimeters (default) or inches. Positive is forward and negative is backward.
 * @param {boolean} [wait=true] will promise wait untill the drive has completed.
 * @returns {Promise}
 */
Hub.prototype.drive = function(distance, wait = true) {
  const angle = Math.abs(distance) * ((this.useMetric ? METRIC_MODIFIER : IMPERIAL_MODIFIER) * this.modifier);
  const dutyCycleA = DRIVE_SPEED * (distance > 0 ? 1 : -1);
  const dutyCycleB = DRIVE_SPEED * (distance > 0 ? 1 : -1);
  return this.motorAngleMultiAsync(angle, dutyCycleA, dutyCycleB, wait);
}

/**
 * Turn robot specified degrees
 * @method Hub#turn 
 * @param {number} degrees degrees to turn. Negative is to the left and positive to the right.
 * @param {boolean} [wait=true] will promise wait untill the turn has completed.
 * @returns {Promise}
 */
Hub.prototype.turn = function(degrees, wait = true) {
  const angle = Math.abs(degrees) * TURN_MODIFIER;
  const dutyCycleA = TURN_SPEED * (degrees > 0 ? 1 : -1);
  const dutyCycleB = TURN_SPEED * (degrees > 0 ? -1 : 1);
  return this.motorAngleMultiAsync(angle, dutyCycleA, dutyCycleB, wait);
}

/**
 * Drive untill sensor shows object in defined distance
 * @method Hub#driveUntil 
 * @param {number} [distance=0] distance in centimeters (default) or inches when to stop. Distance sensor is not very sensitive or accurate.
 * By default will stop when sensor notices wall for the first time. Sensor distance values are usualy between 110-50.
 * @param {boolean} [wait=true] will promise wait untill the bot will stop.
 * @returns {Promise}
 */
Hub.prototype.driveUntil = async function(distance = 0, wait = true) {
  const distanceCheck = distance !== 0 ? (this.useMetric ? distance : distance * 2.54) : DEFAULT_STOP_DISTANCE;
  this.motorTimeMulti(60, DRIVE_SPEED, DRIVE_SPEED);
  if (wait) {
    await waitForValueToSet.bind(this)('distance', () => distanceCheck >= this.distance);
    await this.motorAngleMultiAsync(0);
  }
  else {
    return waitForValueToSet.bind(this)('distance', () => distanceCheck >= this.distance).then(_ => this.motorAngleMulti(0, 0, 0));
  }
}

/**
 * Turn until there is no object in sensors sight
 * @method Hub#turnUntil 
 * @param {number} [direction=1] direction to turn to. 1 (or any positive) is to the right and 0 (or any negative) is to the left.
 * @param {boolean} [wait=true] will promise wait untill the bot will stop.
 * @returns {Promise}
 */
Hub.prototype.turnUntil = async function(direction = 1, wait = true) {
  const directionModifier = direction > 0 ? 1 : -1;
  this.turn(360 * directionModifier, false);
  if (wait) {
    await waitForValueToSet.bind(this)('distance', () => this.distance >= DEFAULT_CLEAR_DISTANCE);
    await this.turn(0, false)  
  }
  else {
    return waitForValueToSet.bind(this)('distance', () => this.distance >= DEFAULT_CLEAR_DISTANCE).then(_ => this.turn(0, false));
  }
}

/**
 * Get BLE status when BLE is ready to be used
 * @method Boost#bleReadyAsync
 * @returns {Promise<boolean>} ble status `true`/`false` when ble is ready
 */
Boost.prototype.bleReadyAsync = function() {
  return new Promise(async (resolve, reject) => {
    var ready = await waitForValueToSet.bind(this)('bleReadyStatus');
    if (ready)
      resolve(ready);
    else 
      reject(ready);
  });
};

/**
 * Get Hub details when hub is found
 * @method Boost#hubFoundAsync
 * @returns {Promise<{uudi: string, address:string, localName: string}>} Hub details
 */
Boost.prototype.hubFoundAsync = function() {
  return waitForValueToSet.bind(this)('hubDetails');
};

/**
 * @method Boost#connectAsync
 * @param hubDetails {object} MAC Address of the Hub
 * @param hubDetails.uuid {string}
 * @param hubDetails.address{string}
 * @param hubDetails.localName {string}
 * @returns {Promise<Hub>} Hub object
 */
Boost.prototype.connectAsync = function(hubDetails) {
  return new Promise((resolve, reject) => {
    this.connect(hubDetails.address, async (err, hub) => {
      if (err) {
        reject(err);
      } else {
        hub.afterInitialization();
        await waitForValueToSet.bind(hub)('connected');
        resolve(hub);
      }
    });
  });
};

/**
 * Connect to a MoveHub and get Hub instance
 * @method Boost#getHubAsync
 * @returns {Promise<Hub>} Hub object 
 */
Boost.prototype.getHubAsync = async function() {
  await this.bleReadyAsync();
  const connectDetails = await this.hubFoundAsync();
  return await this.connectAsync(connectDetails);
};

/**
 * Execute this method after new instance is created
 * @method Boost#afterInitialization
 */
Boost.prototype.afterInitialization = function() {
  this.bleReadyStatus = null;
  this.hubDetails = null;

  this.on('ble-ready', status => (this.bleReadyStatus = status));
  this.on('hub-found', hubDetails => (this.hubDetails = hubDetails));
};

module.exports.Boost = Boost;
module.exports.Hub = Hub;

tester.js
const boost = require('./index');

(async () => {
  // await boost.bleReadyAsync();
  // const connectDetails = await boost.hubFoundAsync();
  // const hub = await boost.connectAsync(connectDetails);

  const hub = await boost.getHubAsync();

  hub.on('error', err => {
    console.log('error', err);
  });

  hub.on('disconnect', () => {
    console.log('disconnect');
  });

  hub.on('distance', distance => {
    console.log('distance', distance);
  });

  hub.on('rssi', rssi => {
    console.log('rssi', rssi);
  });

  hub.on('port', portObject => {
    console.log('port', JSON.stringify(portObject, null, 1));
  });

  hub.on('color', color => {
    console.log('color', color);
  });

  hub.on('tilt', tilt => {
    console.log('tilt', JSON.stringify(tilt, null, 1));
  });

  hub.on('rotation', rotation => {
    console.log('rotation', JSON.stringify(rotation, null, 1));
  });

  // Insert commands here
  await hub.ledAsync('red');
  await hub.ledAsync('yellow');
  await hub.ledAsync('green');

  // Turn 180 degrees
  await hub.motorAngleAsync('A', 980, 100, true);
  
  await hub.motorTimeMultiAsync(2, 100, 100, true);
  
  // Turn 180 degrees  
  await hub.motorAngleAsync('B', 980, 100, true);
  
  await hub.ledAsync('red');  
})();